plugins/devlink/fu-devlink-device.c

```diff
@@ -0,0 +1,828 @@
+
+	if (total > 0)
+		fu_progress_set_percentage_full(helper->progress, done, total);
+
+	return MNL_CB_OK;
+}
+
+/* send flash command */
+static gboolean
+fu_devlink_device_flash_send(FuDevlinkDevice *self,
+			     const gchar *component_name,
+			     const gchar *filename,
+			     GError **error)
+{
+	struct nlmsghdr *nlh;
+
+	/* prepare flash update command */
+	nlh = fu_devlink_netlink_cmd_prepare(self->nlg, DEVLINK_CMD_FLASH_UPDATE, FALSE);
+
+	mnl_attr_put_strz(nlh, DEVLINK_ATTR_BUS_NAME, self->bus_name);
+	mnl_attr_put_strz(nlh, DEVLINK_ATTR_DEV_NAME, self->dev_name);
+
+	if (component_name != NULL) {
+		mnl_attr_put_strz(nlh, DEVLINK_ATTR_FLASH_UPDATE_COMPONENT, component_name);
+		g_debug("sending flash update command for %s/%s with component %s and file %s",
+			self->bus_name,
+			self->dev_name,
+			component_name,
+			filename);
+	} else {
+		g_debug("sending flash update command for %s/%s with file %s",
+			self->bus_name,
+			self->dev_name,
+			filename);
+	}
+
+	mnl_attr_put_strz(nlh, DEVLINK_ATTR_FLASH_UPDATE_FILE_NAME, filename);
+
+	/* send flash update command - this will block until completion */
+	if (!fu_devlink_netlink_msg_send(self->nlg, nlh, error))
+		return FALSE;
+	g_debug("flash update command completed successfully");
+
+	return TRUE;
+}
+
+/* thread function that sends the flash update command and quits main loop */
+static gpointer
+fu_devlink_device_flash_send_thread(gpointer user_data)
+{
+	FuDevlinkFlashSendHelper *helper = user_data;
+	gboolean ret;
+
+	g_debug("flash send thread started for %s/%s",
+		helper->self->bus_name,
+		helper->self->dev_name);
+	ret = fu_devlink_device_flash_send(helper->self,
+					   helper->component_name,
+					   helper->filename,
+					   helper->error);
+
+	/* signal completion by quitting the main loop  */
+	g_main_loop_quit(helper->loop);
+
+	return GINT_TO_POINTER(ret ? 1 : 0);
+}
+
+/* netlink callback for flash progress monitoring */
+static gboolean
+fu_devlink_device_flash_mon_netlink_cb(GIOChannel *channel,
+				       GIOCondition condition,
+				       gpointer user_data)
+{
+	FuDevlinkFlashMonHelper *helper = user_data;
+	gsize len;
+	GIOStatus status;
+	g_autoptr(GError) error = NULL;
+
+	if (condition & (G_IO_ERR | G_IO_HUP)) {
+		g_debug("devlink netlink socket error during flash monitoring");
+		return FALSE;
+	}
+
+	/* read netlink message via GIOChannel */
+	status = g_io_channel_read_chars(channel,
+					 fu_devlink_netlink_gen_socket_get_buf(helper->nlg),
+					 FU_DEVLINK_NETLINK_BUF_SIZE,
+					 &len,
+					 &error);
+	if (status != G_IO_STATUS_NORMAL) {
+		if (error != NULL)
+			g_debug("failed to read devlink netlink message: %s", error->message);
+		return TRUE;
+	}
+
+	/* process netlink messages */
+	if (!fu_devlink_netlink_msg_run(helper->nlg,
+					len,
+					0,
+					fu_devlink_device_flash_mon_cb,
+					helper,
+					&error))
+		g_warning("failed to process netlink message: %s", error->message);
```

Comment:
should this be return FALSE?

Reply:
If I'm reading the documentation correctly, that would stop the monitoring, wouldn't it?

Reply:
done